### PR TITLE
[special] Add special.ndtri

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #50345

TODO
* [x] Add docs https://13865352-65600975-gh.circle-artifacts.com/0/docs/special.html#torch.special.ndtri
* [x] Add comments on implementation
* [x] Clean-up

### CODE REVIEW

**Comment by kshitij12345** [comment]
@mruberry I think this is ready for a review :) Thanks!

**Comment by kshitij12345** [comment]
I did try the composite version. But unfortunately it was failing a few tests at default tolerance (didn't dig deeper into it) also it was slower than kernel.
```
[- ndtri : input dtype torch.float32 device cpu ]
                         |  Composite  |   Kernel
1 threads: --------------------------------------
      (128,)             |      11.1   |      2.0
      (256, 128)         |     496.4   |    227.8
      (2, 512, 256)      |    3928.7   |   1817.3
      (2, 64, 256, 128)  |   66743.5   |  29969.0

Times are in microseconds (us).

[- ndtri : input dtype torch.float64 device cpu ]
                         |  Composite  |   Kernel
1 threads: --------------------------------------
      (128,)             |       10.6  |      1.9
      (256, 128)         |      688.5  |    233.0
      (2, 512, 256)      |     5811.6  |   1865.8
      (2, 64, 256, 128)  |   106655.7  |  37799.4

Times are in microseconds (us).

[- ndtri : input dtype torch.float32 device cuda --]
                             |  Composite  |  Kernel
1 threads: -----------------------------------------
      (128,)                 |      19.9   |     4.4
      (256, 128)             |      20.0   |     5.6
      (2, 512, 256)          |      20.9   |    13.0
      (2, 64, 256, 128)      |     249.7   |   146.7
      (4, 2, 512, 256, 128)  |    7775.1   |  4545.0

Times are in microseconds (us).

[-- ndtri : input dtype torch.float64 device cuda --]
                             |  Composite  |   Kernel
1 threads: ------------------------------------------
      (128,)                 |      19.4   |      5.5
      (256, 128)             |      22.0   |     18.4
      (2, 512, 256)          |      68.0   |     74.6
      (2, 64, 256, 128)      |    1059.9   |    954.3
      (4, 2, 512, 256, 128)  |   33449.8   |  29739.4

Times are in microseconds (us).
```

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

**Comment by mruberry** [comment]
> @mruberry gentle ping :)

Thanks for the pings! Sorry for the delay

File: aten/src/ATen/native/Math.h [hunk 1]
```diff
@@ -220,16 +220,9 @@ static inline double zeta(double x, double q) {
...
-  double result = 0;
-  for (size_t i = 0; i <= len; i++) {
-    result = result * x + A[i];
-  }
-  return result;
-}
-
-static inline float polevlf(float x, float *A, size_t len) {
-  float result = 0;
+template <typename T>
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:223]: Would you add a comment for this function? How to use it and what it's intended for?

File: aten/src/ATen/native/Distributions.h [hunk 2]
```diff
@@ -118,15 +119,6 @@ C10_DEVICE scalar_t sample_gamma(scalar_t alpha, BaseSampler<accscalar_t, unifor
   }
 }
 
-template <typename scalar_t>
```
Comment (mruberry) [inline aten/src/ATen/native/Distributions.h:121]: Nice simplification

File: aten/src/ATen/native/Math.h [hunk 3]
```diff
@@ -384,7 +377,7 @@ static inline float calc_digamma(float x) {
   float y = 0;
   if (x < 1.0e17f) {
     float z = 1 / (x * x);
-    y = z * polevlf(z, A, 6);
+    y = z * polevl(z, A, 6);
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:380]: What does this change do?

File: aten/src/ATen/native/Math.h [hunk 4]
```diff
@@ -1247,7 +1240,7 @@ inline std::tuple<const T*, size_t> chebyshev_coefficients_i0e_B() {
 };
 
 template <typename T>
-inline typename std::enable_if<std::is_same<double, T>::value, std::tuple<const T*, size_t>>::type
+static inline typename std::enable_if<std::is_same<double, T>::value, std::tuple<const T*, size_t>>::type
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:1243]: What does declaring these static do?

File: aten/src/ATen/native/Math.h [hunk 5]
```diff
@@ -1461,3 +1454,135 @@ calc_i1e(T _x) {
       static_cast<T>(chbevl(static_cast<T>(32.0 / x - 2.0), B, len) / std::sqrt(x));
   return (_x < 0.0) ? -out : out;
 }
+
+/*
+ * This function is derived from the implementation of the i1e function in the Cephes Math Library.
```
Comment (mruberry) [inline aten/src/ATen/native/Math.h:1459]: Nice comment

**Review by mruberry — APPROVED** [review]
Looks perfect. Stellar work, @kshitij12345!

File: aten/src/ATen/native/Math.h (hunk 3 above)
Comment (kshitij12345) [reply aten/src/ATen/native/Math.h:380]: Previously this file had `double polevl(double x, double *A, size_t len)` for double and `float polevlf(float x, float *A, size_t len)` for floating inputs. We have no replaced them with a template `polevl`. Hence the change.

File: aten/src/ATen/native/Math.h (hunk 4 above)
Comment (kshitij12345) [reply aten/src/ATen/native/Math.h:1243]: This header is included in multiple files. With static, we are assuring the compiler that this function will only be used in current translation unit. Without static,(I think) compiler is not free to remove the code even if it is unused as some other translation unit may refer to it.

Comment (kshitij12345) [reply aten/src/ATen/native/Math.h:1243]: Without static, the assembly still has unused function

https://godbolt.org/z/Td8exxYMP

With static, it is removed

https://godbolt.org/z/Y7v5EMh6h

**Comment by kshitij12345** [comment]
Had to resolve a conflict. (might have to pull this again)

**Pull request closed without merging**